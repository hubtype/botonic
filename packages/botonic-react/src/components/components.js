/**
 * Get the name of an element (instantiated component)
 * @param {React.Element} component
 * @returns {string}
 */
export function getElementName(component) {
  return component.type.name
}

/**
 * Get the name of a component TYPE
 * @param {React.ComponentType} component
 * @returns {string}
 */
export function getComponentTypeName(component) {
  return component.displayName || component.name || 'Component'
}
