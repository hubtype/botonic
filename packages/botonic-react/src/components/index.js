export { Audio } from './audio'
export { Button } from './button'
export { Carousel } from './carousel'
export { customMessage } from './custom-message'
export { Document } from './document'
export { Element } from './element'
export { Image } from './image'
export { Location } from './location'
export { MessageTemplate } from './message-template'
export * from './multichannel'
export { Pic } from './pic'
export { Reply } from './reply'
export { ShareButton } from './share-button'
export { Subtitle } from './subtitle'
export { Text } from './text'
export { Title } from './title'
export { Video } from './video'
export { WebchatSettings } from './webchat-settings'
export { WhatsappTemplate } from './whatsapp-template'
