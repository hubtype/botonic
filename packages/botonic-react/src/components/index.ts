export { Audio } from './audio'
export { Button } from './button'
export { Carousel } from './carousel'
export { customMessage } from './custom-message'
export { Document } from './document'
export { Element } from './element'
export { Image } from './image'
export * from './index-types'
export { Location } from './location'
export { Message } from './message'
export { MessageTemplate } from './message-template'
export * from './multichannel'
export { Pic } from './pic'
export { Raw } from './raw'
export { Reply } from './reply'
export { ShareButton } from './share-button'
export { Subtitle } from './subtitle'
export { Text } from './text'
export { Title } from './title'
export { Video } from './video'
export { WebchatSettings } from './webchat-settings'
export {
  WhatsappButtonList,
  WhatsappButtonListProps,
  WhatsappButtonListRowProps,
  WhatsappButtonListSectionProps,
} from './whatsapp-button-list'
export {
  WhatsappCTAUrlButton,
  WhatsappCTAUrlButtonProps,
} from './whatsapp-cta-url-button'
export { WhatsappTemplate } from './whatsapp-template'
