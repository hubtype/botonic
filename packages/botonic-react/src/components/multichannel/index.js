import { Multichannel } from './multichannel'
import { MultichannelText } from './m-text'
import { MultichannelButton } from './m-button'
import { MultichannelReply } from './m-reply'
import { MultichannelCarousel } from './m-carousel'

export { Multichannel }
export { MultichannelText }
export { MultichannelButton }
export { MultichannelReply }
export { MultichannelCarousel }
