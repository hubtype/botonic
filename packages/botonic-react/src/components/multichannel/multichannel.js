import React from 'react'
import { RequestContext } from '../../contexts'

export class Multichannel extends React.Component {
  static contextType = RequestContext
  constructor(props) {
    super(props)
  }
  render() {
    if (this.context.session.user.provider == this.props.channel) {
      return this.props.children
    }
    return null
  }
}
