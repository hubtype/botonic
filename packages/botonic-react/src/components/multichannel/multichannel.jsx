import React, { useContext } from 'react'

import { COMPONENT_TYPE } from '../../constants'
import { RequestContext } from '../../contexts'
import { deepMapWithIndex } from '../../util/react'
import { Text } from '../text'
import { MultichannelButton } from './multichannel-button'
import { MultichannelCarousel } from './multichannel-carousel'
import { MultichannelContext } from './multichannel-context'
import { MultichannelReply } from './multichannel-reply'
import { MultichannelText } from './multichannel-text'
import {
  isFacebook,
  isWhatsapp,
  MULTICHANNEL_WHATSAPP_PROPS,
} from './multichannel-utils'

export const Multichannel = props => {
  const requestContext = useContext(RequestContext)

  if (!isWhatsapp(requestContext) && !isFacebook(requestContext)) {
    return props.children
  }
  if (isFacebook(requestContext)) {
    const newChildren = deepMapWithIndex(props.children, child => {
      if (child && child.type && child.type.name === COMPONENT_TYPE.TEXT) {
        return (
          <MultichannelText {...child.props} key={child.key}>
            {child.props.children}
          </MultichannelText>
        )
      }
      return child
    })
    return newChildren
  }

  let newChildren = deepMapWithIndex(props.children, (child, index) => {
    if (child && child.type && child.type.name === COMPONENT_TYPE.BUTTON) {
      return (
        <MultichannelButton {...child.props} key={child.key}>
          {child.props.children}
        </MultichannelButton>
      )
    }
    if (child && child.type && child.type.name === COMPONENT_TYPE.REPLY) {
      return (
        <MultichannelReply {...child.props} key={child.key}>
          {child.props.children}
        </MultichannelReply>
      )
    }
    if (child && child.type && child.type.name === COMPONENT_TYPE.TEXT) {
      return (
        <MultichannelText
          {...child.props}
          {...props.text}
          key={child.key}
          {...(props.messageSeparator &&
            index > 0 && { newline: props.messageSeparator })}
        >
          {child.props.children}
        </MultichannelText>
      )
    }
    if (child && child.type && child.type.name === COMPONENT_TYPE.CAROUSEL) {
      return (
        <MultichannelCarousel
          {...child.props}
          {...props.carousel}
          key={child.key}
        >
          {child.props.children}
        </MultichannelCarousel>
      )
    }
    return child
  })
  if (props.messageSeparator != null) {
    newChildren = newChildren.map((c, index) =>
      index > 0 && typeof c === 'string' ? props.messageSeparator + c : c
    )
    newChildren = (
      <Text key={props.key} {...MULTICHANNEL_WHATSAPP_PROPS}>
        {newChildren}
      </Text>
    )
  }
  return (
    <MultichannelContext.Provider
      value={{
        currentIndex: props.firstIndex,
        boldIndex: props.boldIndex,
        indexSeparator: props.indexSeparator,
        messageSeparator: props.messageSeparator,
      }}
    >
      {newChildren}
    </MultichannelContext.Provider>
  )
}
