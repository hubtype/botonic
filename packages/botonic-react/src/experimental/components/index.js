export { Button } from '../../components/button'
export { Element } from '../../components/element'
export { MessageTemplate } from '../../components/message-template'
export * from '../../components/multichannel'
export { Pic } from '../../components/pic'
export { Reply } from '../../components/reply'
export { ShareButton } from '../../components/share-button'
export { Subtitle } from '../../components/subtitle'
export { Title } from '../../components/title'
export { WebchatSettings } from '../../components/webchat-settings'

// Experimental
export { Audio } from './audio'
export { Carousel } from './carousel'
export { customMessage } from './custom-message'
export { Document } from './document'
export { Image } from './image'
export { Location } from './location'
export { Text } from './text'
export { Video } from './video'
export { WhatsappTemplate } from './whatsapp-template'
