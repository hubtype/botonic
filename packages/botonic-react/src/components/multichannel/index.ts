import { Multichannel } from './multichannel'
import { MultichannelButton } from './multichannel-button'
import { MultichannelCarousel } from './multichannel-carousel'
import { MultichannelReply } from './multichannel-reply'
import { MultichannelText } from './multichannel-text'

export { Multichannel }
export { MultichannelText }
export { MultichannelButton }
export { MultichannelReply }
export { MultichannelCarousel }
