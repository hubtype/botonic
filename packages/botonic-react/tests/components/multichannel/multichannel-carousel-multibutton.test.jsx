import { Button, Carousel, Element, Pic, Subtitle, Title } from '../../../src'
import { MultichannelCarousel } from '../../../src/components/multichannel'
import React from 'react'
import { whatsappRenderer } from '../../helpers/test-utils'

let movies = [
  {
    name: 'Snatch',
    desc: 'Five minutes, Turkish',
    url: 'https://www.imdb.com/title/tt0208092',
    pic:
      'https://nebula.wsimg.com/obj/NzQ3QUYxQzZBNzE4NjNFRTc1MTU6NmM4YjgzZWVlZTE2MGMzM2RkMTdlZjdjNGUyZmFhNDE6Ojo6OjA=',
  },
  {
    name: 'El laberinto del fauno',
    desc: '2006',
    url: 'https://www.imdb.com/title/tt0457430',
    pic:
      'https://m.media-amazon.com/images/M/MV5BMDBjOWYyMDQtOWRmOC00MDgxLWIxM2UtMTNjYzFiN2RkYTBlXkEyXkFqcGdeQXVyMTAyOTE2ODg0._V1_UY268_CR6,0,182,268_AL_.jpg',
  },
]

describe('Multichannel carousel COMPACT mode N Buttons', () => {
  test('dynamic carousel with url buttons', () => {
    const carousel = (
      <Carousel>
        {movies.map((e, i) => (
          <Element key={e.name}>
            <Pic src={e.pic} />
            <Title>{e.name}</Title>
            <Subtitle>{e.desc}</Subtitle>
            {[
              <Button key={'1'} url={e.url}>
                Visit website
              </Button>,
              <Button key={'2'} url={'anotherurl.com'}>
                Another website
              </Button>,
            ]}
          </Element>
        ))}
      </Carousel>
    )
    const sut = (
      <MultichannelCarousel {...carousel.props}>
        {carousel.props.children}
      </MultichannelCarousel>
    )
    const renderer = whatsappRenderer(sut)
    const tree = renderer.toJSON()
    expect(tree).toMatchSnapshot()
  })

  const options = [
    {
      name: 'Title1',
      desc: 'Subtitle1',
      payload: 'Payload1',
      buttonText: 'Previo a la compra',
      pic:
        'https://cdn.slidesharecdn.com/profile-photo-sumeet.moghe-48x48.jpg?cb=1527568614',
    },
    {
      name: 'Title2',
      desc: 'Subtitle2',
      payload: 'Payload2',
      buttonText: 'Durante la compra',
      pic:
        'https://nebula.wsimg.com/obj/NzQ3QUYxQzZBNzE4NjNFRTc1MTU6NmM4YjgzZWVlZTE2MGMzM2RkMTdlZjdjNGUyZmFhNDE6Ojo6OjA=',
    },
  ]

  test('dynamic carousel with payload/path buttons N Buttons', () => {
    const carousel = (
      <Carousel>
        {options.map((e, i) => (
          <Element key={e.name}>
            <Pic src={e.pic} />
            <Title>{e.name}</Title>
            <Subtitle>{e.desc}</Subtitle>
            {[
              <Button key={'1'} payload={'payload1'}>
                {e.buttonText}
              </Button>,
              <Button key={'2'} payload={'payload2'}>
                another button{i}
              </Button>,
            ]}
          </Element>
        ))}
      </Carousel>
    )
    const sut = (
      <MultichannelCarousel {...carousel.props}>
        {carousel.props.children}
      </MultichannelCarousel>
    )
    const renderer = whatsappRenderer(sut)
    const tree = renderer.toJSON()
    expect(tree).toMatchSnapshot()
  })

  const optionsNoTitleNoSubtitle = [
    {
      name: '',
      desc: '',
      payload: 'Payload1',
      buttonText: 'Previo a la compra',
      pic:
        'https://cdn.slidesharecdn.com/profile-photo-sumeet.moghe-48x48.jpg?cb=1527568614',
    },
    {
      name: '',
      desc: '',
      payload: 'Payload2',
      buttonText: 'Durante la compra',
      pic:
        'https://nebula.wsimg.com/obj/NzQ3QUYxQzZBNzE4NjNFRTc1MTU6NmM4YjgzZWVlZTE2MGMzM2RkMTdlZjdjNGUyZmFhNDE6Ojo6OjA=',
    },
  ]

  test('dynamic carousel (just text in button) with payload/path buttons N Buttons', () => {
    const carousel = (
      <Carousel>
        {optionsNoTitleNoSubtitle.map((e, i) => (
          <Element key={e.name}>
            <Pic src={e.pic} />
            <Title>{e.name}</Title>
            <Subtitle>{e.desc}</Subtitle>
            {[
              <Button key={'1'} payload={e.payload}>
                {e.buttonText}
              </Button>,
              <Button key={'2'} payload={e.payload}>
                another button{i}
              </Button>,
            ]}
          </Element>
        ))}
      </Carousel>
    )
    const sut = (
      <MultichannelCarousel {...carousel.props}>
        {carousel.props.children}
      </MultichannelCarousel>
    )
    const renderer = whatsappRenderer(sut)
    const tree = renderer.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
