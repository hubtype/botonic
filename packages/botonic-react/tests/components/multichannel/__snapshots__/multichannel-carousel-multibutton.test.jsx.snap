// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Multichannel carousel COMPACT mode N Buttons dynamic carousel (just text in button) with payload/path buttons N Buttons 1`] = `
Array [
  <message
    delay={0}
    indexMode="number"
    markdown={false}
    newkey={0}
    type="text"
    typing={0}
  >
    1. Previo a la compra
    
2. another button,0
  </message>,
  <message
    delay={0}
    indexMode="number"
    markdown={false}
    newkey={1}
    type="text"
    typing={0}
  >
    3. Durante la compra
    
4. another button,1
  </message>,
]
`;

exports[`Multichannel carousel COMPACT mode N Buttons dynamic carousel with payload/path buttons N Buttons 1`] = `
Array [
  <message
    delay={0}
    indexMode="number"
    markdown={false}
    newkey={0}
    type="text"
    typing={0}
  >
    *Title1* _Subtitle1_
    
1. Previo a la compra
    
2. another button,0
  </message>,
  <message
    delay={0}
    indexMode="number"
    markdown={false}
    newkey={1}
    type="text"
    typing={0}
  >
    *Title2* _Subtitle2_
    
3. Durante la compra
    
4. another button,1
  </message>,
]
`;

exports[`Multichannel carousel COMPACT mode N Buttons dynamic carousel with url buttons 1`] = `
Array [
  <message
    delay={0}
    indexMode="number"
    markdown={false}
    newkey={0}
    type="text"
    typing={0}
  >
    *Snatch* _Five minutes, Turkish_
    
- Visit website: https://www.imdb.com/title/tt0208092
    
- Another website: anotherurl.com
  </message>,
  <message
    delay={0}
    indexMode="number"
    markdown={false}
    newkey={1}
    type="text"
    typing={0}
  >
    *El laberinto del fauno* _2006_
    
- Visit website: https://www.imdb.com/title/tt0457430
    
- Another website: anotherurl.com
  </message>,
]
`;
